p2p/protocol/identify/peer_record.go

```diff
@@ -0,0 +1,172 @@
+// be shared with other peers to inform them of our listen addresses in
+// a secure and authenticated way.
+//
+// New signed records are created in response to EvtLocalAddressesUpdated events,
+// and are emitted in EvtLocalPeerRecordUpdated events.
+//
+// EvtLocalPeerRecordUpdated is emitted using a Stateful emitter, so new subscribers
+// will immediately receive the current record when they subscribe, with future
+// records delivered in future events.
+type peerRecordManager struct {
+	lock       sync.RWMutex
+	latest     *record.Envelope
+	hostID     peer.ID
+	signingKey crypto.PrivKey
+
+	ctx           context.Context
+	subscriptions struct {
+		localAddrsUpdated event.Subscription
+	}
+	emitters struct {
+		evtLocalPeerRecordUpdated event.Emitter
+	}
+}
+
+// NewPeerRecordManager creates a peerRecordManager that will subscribe to the given event.Bus
+// and listen for changes in the local Host's addresses, emitting new signed peer.PeerRecords
+// in response. The new records will be contained in event.EvtLocalPeerRecordUpdated events
+// and emitted on the event bus.
+//
+// PeerRecords will be signed with the given private key, which must be the libp2p Host's
+// identity key for the resulting records to be valid.
+//
+// If initialAddrs is non-empty, a PeerRecord will be created immediately and emitted on
+// the bus, without waiting for an event.LocalPeerAddressesUpdated event to trigger an
+// update.
+func NewPeerRecordManager(ctx context.Context, bus event.Bus, hostKey crypto.PrivKey, initialAddrs []multiaddr.Multiaddr) (*peerRecordManager, error) {
+	hostID, err := peer.IDFromPrivateKey(hostKey)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &peerRecordManager{
+		ctx:        ctx,
+		signingKey: hostKey,
+		hostID:     hostID,
+	}
+
+	if len(initialAddrs) != 0 {
+		initialRec, err := m.makeSignedPeerRecord(initialAddrs)
+		if err != nil {
+			return nil, fmt.Errorf("error constructing initial peer record: %w", err)
+		}
+		m.latest = initialRec
+	}
+
+	m.subscriptions.localAddrsUpdated, err = bus.Subscribe(&event.EvtLocalAddressesUpdated{}, eventbus.BufSize(128))
+	if err != nil {
+		return nil, err
+	}
+	m.emitters.evtLocalPeerRecordUpdated, err = bus.Emitter(&event.EvtLocalPeerRecordUpdated{}, eventbus.Stateful)
+	if err != nil {
+		return nil, err
+	}
+
+	if m.latest != nil {
+		m.emitLatest()
+	}
+
+	go m.handleEvents()
+
+	return m, nil
+}
+
+// LatestRecord returns the most recently constructed signed PeerRecord.
+// If you have a direct reference to the peerRecordManager, this method
+// gives you a simple way to access the latest record without pulling
+// from the event bus.
+func (m *peerRecordManager) LatestRecord() *record.Envelope {
+	if m == nil {
+		return nil
+	}
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	return m.latest
+}
+
+func (m *peerRecordManager) handleEvents() {
+	sub := m.subscriptions.localAddrsUpdated
+	defer func() {
+		_ = sub.Close()
+		// drain the channel.
+		for range sub.Out() {
+		}
```

Comment:
@raulk Is this required by the API? That's kind of dangerous.

Reply:
It's superfluous here as go-eventbus already drains the channel on close:

https://github.com/libp2p/go-eventbus/blob/master/basic.go#L110-L115

We should document this behaviour in the `Subscription` interface.

Reply:
Ah, that makes more sense. Then we should probably remove this.